Adding South Africa

yukinko-iwasaki opened the pull request:
- Added South Africa
- Minor fix for Albania

The subnational population is added here.
https://github.com/dime-worldbank/mega-indicators/pull/9

weilu commented on cross_country_aggregate_dlt.py:
```diff
@@ -25,6 +25,7 @@
     StructField("executed", DoubleType(), True, {'comment': 'Actual amount spent in current local currency by the end of the fiscal year.'})
 ])
 
+@dlt.expect_or_drop("executed not null nor 0", "executed IS NOT NULL and executed != 0")
```
Update it to drop if both executed and approved are null or 0

yukinko-iwasaki replied:
done!

yukinko-iwasaki commented on South_Africa/ZAF_transform_load_dlt.py:
```diff
@@ -0,0 +1,404 @@
+    for old_col_name in bronze_df.columns:
+        new_col_name = old_col_name.replace(" ", "_")
+        bronze_df = bronze_df.withColumnRenamed(old_col_name, new_col_name).withColumn(
+            "id", monotonically_increasing_id()
```
oh..  I don't think we need to keep this column. It was for debugging purpose only.

yukinko-iwasaki commented on South_Africa/ZAF_transform_load_dlt.py:
```diff
@@ -0,0 +1,404 @@
+                (col("Economic_Level_1") != "Payments for financial assets")
+                & (col("Economic_Level_2") != "Provinces and municipalities")
+            )
+            .withColumn("econ1", coalesce(col("Economic_Level_1"), lit("")))
+            .withColumn("econ2", coalesce(col("Economic_Level_2"), lit("")))
+            .withColumn("econ3", coalesce(col("Economic_Level_3"), lit("")))
+            .withColumn("func_lower", coalesce(lower(col("Function_group")), lit("")))
+            .withColumn("func_sub_lower", coalesce(lower(col("Budget_group")), lit("")))
+            .withColumn("admin0", coalesce(col("Admin1"), lit("")))
+            .withColumn("admin0_temp", coalesce(col("Admin1"), lit("")))
+            .withColumn("admin1", coalesce(col("Department"), lit("")))
```
hmm.. 
I think fillna should be enough, but for readability, coalesce might be better. 
fillna does not let up work on individual column. It has to be called on top of the whole df and specify the subset of column names to apply the changes.

yukinko-iwasaki commented on South_Africa/ZAF_transform_load_dlt.py:
```diff
@@ -0,0 +1,404 @@
+            .when(
+                (
+                    (
+                        col("func_sub_lower").isin(
+                            "basic education",
+                            "education",
+                        )
```
no formula available for tertiary education.

yukinko-iwasaki commented on South_Africa/ZAF_transform_load_dlt.py:
```diff
@@ -0,0 +1,404 @@
+                "Regional",
+            ).when((col("admin0_temp") == "Central"), "Central"),
+        )
+        .withColumn(
+            "admin1",
+            when(col("admin0") == "Central", "Central Scope").when(
+                col("admin0") == "Regional", col("admin1")
+            ),
+        )
+        .withColumn(
+            "geo1",
+            when(col("admin0").isNull(), "Central Scope")
+            .when(
+                col("admin0") == "Central",
+                "Central Scope",
+            )
+            .otherwise(
+                col("admin1"),
+            ),
+        )
```
You are right. admin1 and geo1 are the same. I will use the alias for this case. 
Admin1 is the excel sheet refers to our admin0 while departmnent is equivalent of our admin1

yukinko-iwasaki commented:
@weilu thanks for the review.
I think I addressed your comments + addressed the outstanding discrepancy issues.

I found further causes of discrepancies and had to add extra exclusions to social assisstence with the following additional tags : Wage bill, Goods and services, Other grants and transfers, Capital Expenditures

More details here:
https://docs.google.com/document/d/10wdHD5x2IYw6VC-savb2TcC7y1adIPle2dMKOE0D7sI/edit?tab=t.0#heading=h.vpzw9mxhtplx

I will send an email to Massimo regarding missing countries and this new update on South Africa tomorrow afternoon.

yukinko-iwasaki commented:
@weilu thanks for the review.
I think I addressed your comments + addressed the outstanding discrepancy issues.

I found further causes of discrepancies and had to add extra exclusions to social assisstence with the following additional tags : Wage bill, Goods and services, Other grants and transfers, Capital Expenditures

More details here in Massimo's question doc. 

I will send an email to Massimo regarding missing countries and this new update on South Africa tomorrow afternoon.

weilu commented on South_Africa/ZAF_transform_load_dlt.py:
```diff
@@ -80,28 +73,17 @@ def boost_silver():
-            when(col("admin0").isNull(), "Central Scope")
-            .when(
-                col("admin0") == "Central",
+            when(
+                col("admin0_temp").isin("Provinces", "Municipalities"),
```
Consider `when(col('admin0') == 'Regional', col('Department'))` to avoid repeating the logic.

weilu reviewed (approved):
@yukinko-iwasaki LGTM with one minor comment for your consideration.
